stubs/serverless.yml

```diff
@@ -0,0 +1,46 @@
+service: laravel
+
+provider:
+    name: aws
+    # The AWS region in which to deploy (us-east-1 is the default)
+    region: us-east-1
+    # The stage of the application, e.g. dev, production, staging… ('dev' is the default)
+    stage: dev
+    runtime: provided.al2
+
+package:
+    # Directories to exclude from deployment
+    patterns:
+        - '!node_modules/**'
+        - '!public/storage'
+        - '!resources/assets/**'
+        - '!storage/**'
+        - '!tests/**'
+
+functions:
+
+    # This function runs the Laravel website/API
+    web:
+        handler: public/index.php
+        timeout: 28 # in seconds (API Gateway has a timeout of 29 seconds)
+        layers:
+            - ${bref:layer.php-81-fpm}
+        events:
+            - httpApi: '*'
+
+    # This function lets us run artisan commands in Lambda
+    artisan:
+        handler: artisan
```

Comment:
I think that this addresses https://github.com/brefphp/laravel-bridge/discussions/79#discussioncomment-4177673 but not 100%

I.e. the initialization code in the custom runtime isn't executed anymore (and actually that's the case for the `web` function as well.

I probably need to address https://github.com/brefphp/laravel-bridge/discussions/79#discussioncomment-4177673 then 🤔

Reply:
I have pushed a new commit (https://github.com/cachewerk/bref-laravel-bridge/pull/36/commits/b00151cb4d6cbf6cbe863494c741aaa0d4738817) that moves all the initialization logic into the BrefServiceProvider.

It seems to work, but I haven't tested with SSM parameters. Having automated tests would also really help increase confidence with these changes too.

But I'm submitting the code for review here to get your opinion first. I like the change because it removes "custom" solutions specific to Bref (i.e. creating a "runtime" file). It becomes more "Laravel native", since all the logic is in the service provider.

I think what we need next is also to handle cases where the app is running locally (not on Lambda) so that all the workarounds are not applied (e.g. don't cache the config in dev env). But that can come in a future PR I think (because the problem already exists on master).

Reply:
Have you tested this? I think Laravel loads the config way before the `BrefServiceProvider`.

Have you tested whether the Octane implementation still works with all this moved to the service provider?

Reply:
Yeah I'm not too sure either.

I'm trying to take a step back and think more holistically about this. What if the loading of encrypted env variables was a native Bref feature that worked not just for Laravel? After all, there isn't anything specific to Laravel in there. This could be a Bref 2.0 feature.

I also looked at https://github.com/customink/crypteia and like their approach (though I don't feel comfortable reusing their solution yet, it's still too early and not proven enough). It's generic enough and mimics existing solutions well.

If we move the "decrypt environment variables" logic to Bref, then what's left in the "boot" of the runtime is:

- create storage directories
- `$app->useStoragePath()`
- run `artisan config:cache` once

I _think_ all these can be put in BrefServiceProvider::boot(), but what about the last one? Do you think that will work out?

Reply:
I experimented with this feature in Bref, here is a PR: https://github.com/brefphp/bref/pull/1376

Reply:
https://github.com/cachewerk/bref-laravel-bridge/pull/29

Reply:
Makes sense to bake this all into bref 2. Ideally there should be some bootstrap script that one can extend upfrom from userland. This will come in handy when you want to run caching steps etc during init (especially when/if SnapStart comes available for custom runtimes).

Reply:
`LoadEnvironmentVariables` is executed way before service provider's `boot()` or `register()` is called, so no this won't work.

Reply:
The storage path must be set prior to bootstrapping Laravel, that won't work in a SP either.

Reply:
Supporting what @georgeboot did in #29 directly in Bref, would be rad of course.

Reply:
> https://github.com/cachewerk/bref-laravel-bridge/pull/29

I don't see how using the extension is better, or how it could be faster. We would issue 1 call to SSM in the same process. The extension would issue one call to SSM in a separate process. Both are very close, but the first is faster and simpler, am I missing something? I don't trust AWS for shipping anything very optimized unless it's a major service 😅

> `LoadEnvironmentVariables` is executed way before service provider's `boot()` or `register()` is called, so no this won't work, so the caching and variable injection has to happen prior.

With what I shared above there is no env var set in the service provider anymore, check out the PR I linked. This all happens before Laravel boots.

> Ideally there should be some bootstrap script that one can extend upfrom from userland. Default is injecting env etc.
> 
> This will come in handy when you want to run caching steps etc during init (especially when/if SnapStart comes available for custom runtimes).

I'm thinking about this, but I don't find an elegant solution for this that:

- would work with all runtimes
- but also would let you run code in the same process as the userland code (FPM is a separate process for example)
- and that also lets users use the feature

Let's me expand on the last item because maybe it's not clear. Let's say we add in Bref a `BREF_BOOTSTRAP` env var: if you set a script in that env var, then it runs on cold start. The Laravel bridge could use it, but then Laravel users cannot use that variable because it's already being used. It's not a good API because it can only be used once.

That's why I'm looking at avoiding that question entirely, and it seems we are really close to it. The last problematic item seems to be `$app->useStoragePath()`.

I'll dig into that and see if there's any way we can make that work.

Reply:
Yeah, if Bref can handle Parameter Store and/or Secrets Manager that would be gold ❤️‍🔥

Reply:
It seems like `useStoragePath()` can be used later in the app boot process (later than `bootstrap/app.php`):

- https://github.com/laravel/vapor-core/blob/9632ea0ad1f59287df456e7fe67968f81bd5b5e5/src/Runtime/HttpKernel.php#L49
- https://github.com/laravel/vapor-core/blob/4a9bae3c1909fcaf7e1de2c3d7bb7f7abe05e9f1/src/Runtime/Octane/Octane.php#L143
- https://github.com/cachewerk/bref-laravel-bridge/blob/main/src/Octane/OctaneClient.php#L59
- https://github.com/brefphp/laravel-bridge/pull/9/files#diff-6888bb7b8ac364b842752c0f4d23a2e669b8dd5ad00f1aca21016df4f38c3779R74

Also in my preliminary tests it worked, though it was a simple app.

Reply:
We set it early, because of the `config:cache` call.